pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockBuilder.java

```diff
@@ -187,200 +141,315 @@ public static RowDataBlock buildFromRows(List<Object[]> rows, DataSchema dataSch
           }
-          break;
-        case STRING_ARRAY:
-          for (int rowId = 0; rowId < numRows; rowId++) {
-            value = column[rowId];
-            if (value == null) {
-              nullBitmaps[colId].add(rowId);
-              value = nullPlaceholders[colId];
-            }
-            setColumn(columnarBuilder, byteBuffer, (String[]) value);
+        }
+        break;
+      }
+      case STRING_ARRAY: {
+        String[] nullPlaceholder = (String[]) storedType.getNullPlaceholder();
+        for (int rowId = 0; rowId < numRows; rowId++) {
+          Object value = column[rowId];
+          if (value == null) {
+            nullBitmap.add(rowId);
+            setColumn(fixedSize, varSize, nullPlaceholder, dictionary);
+          } else {
+            setColumn(fixedSize, varSize, (String[]) value, dictionary);
           }
-          break;
+        }
+        break;
+      }
 
-        // Special intermediate result for aggregation function
-        case OBJECT:
-          for (int rowId = 0; rowId < numRows; rowId++) {
-            setColumn(columnarBuilder, byteBuffer, column[rowId]);
-          }
-          break;
+      // Special intermediate result for aggregation function
+      case OBJECT: {
+        for (int rowId = 0; rowId < numRows; rowId++) {
+          setColumn(fixedSize, varSize, column[rowId]);
+        }
+        break;
+      }
+      // Null
+      case UNKNOWN:
+        for (int rowId = 0; rowId < numRows; rowId++) {
+          setColumn(fixedSize, varSize, (Object) null);
+        }
+        break;
 
-        // Null
-        case UNKNOWN:
-          for (int rowId = 0; rowId < numRows; rowId++) {
-            setColumn(columnarBuilder, byteBuffer, (Object) null);
-          }
-          break;
+      default:
+        throw new IllegalStateException(
+            String.format("Unsupported stored type: %s for column: %s", storedType,
+                dataSchema.getColumnName(colId)));
+    }
+  }
 
+  private static int calculateBytesPerRow(DataSchema dataSchema) {
+    int rowSizeInBytes = 0;
+    for (ColumnDataType columnDataType : dataSchema.getColumnDataTypes()) {
+      switch (columnDataType) {
+        case INT:
+          rowSizeInBytes += 4;
+          break;
+        case LONG:
+          rowSizeInBytes += 8;
+          break;
+        case FLOAT:
+          rowSizeInBytes += 4;
+          break;
+        case DOUBLE:
+          rowSizeInBytes += 8;
+          break;
+        case STRING:
+          rowSizeInBytes += 4;
+          break;
+        // Object and array. (POSITION|LENGTH)
         default:
-          throw new IllegalStateException(
-              String.format("Unsupported stored type: %s for column: %s", storedTypes[colId],
-                  dataSchema.getColumnName(colId)));
+          rowSizeInBytes += 8;
+          break;
       }
-      columnarBuilder._fixedSizeDataByteArrayOutputStream.write(byteBuffer.array(), 0, byteBuffer.position());
```

Comment:
Here again we were copying bytes. In this case the allocation is smaller, but problematic anyway.